src/clients/api.ts

```diff
@@ -0,0 +1,235 @@
+  ResponseDecoder
+} from "italia-ts-commons/lib/requests";
+import nodeFetch from "node-fetch";
+import { ExtendedProfile } from "../types/api/ExtendedProfile";
+import { LimitedProfile } from "../types/api/LimitedProfile";
+import { MessageResponseWithContent } from "../types/api/MessageResponseWithContent";
+import { Messages } from "../types/api/Messages";
+import { ServicePublic } from "../types/api/ServicePublic";
+import { Services } from "../types/api/Services";
+
+const OcpApimSubscriptionKey = "Ocp-Apim-Subscription-Key";
+type OcpApimSubscriptionKey = t.TypeOf<typeof OcpApimSubscriptionKey>;
+
+// ProfileLimitedOrExtended is oneOf [LimitedProfile, ExtendedProfile]
+const ProfileLimitedOrExtended = t.union([LimitedProfile, ExtendedProfile]);
+
+export type ProfileLimitedOrExtended = t.TypeOf<
+  typeof ProfileLimitedOrExtended
+>;
+
+export type BasicResponseTypeWith401<R> =
+  | BasicResponseType<R>
+  | IResponseType<401, Error>;
+
+// A basic response decoder that also include 401
+export function basicResponseDecoderWith401<R, O = R>(
+  type: t.Type<R, O>
+): ResponseDecoder<BasicResponseTypeWith401<R>> {
+  return composeResponseDecoders(
+    basicResponseDecoder(type),
+    basicErrorResponseDecoder(401)
+  );
+}
+
+export function SubscriptionKeyHeaderProducer<P>(
+  token: string
+): RequestHeaderProducer<P, OcpApimSubscriptionKey> {
+  return () => ({
+    [OcpApimSubscriptionKey]: token
+  });
+}
+
+export type GetProfileT = IGetApiRequestType<
+  {
+    readonly fiscalCode: string;
+  },
+  OcpApimSubscriptionKey,
+  never,
+  BasicResponseTypeWith401<ProfileLimitedOrExtended>
+>;
+
+export type CreateOrUpdateProfileT = IPostApiRequestType<
+  {
+    readonly fiscalCode: string;
+    readonly newProfile: ExtendedProfile;
+  },
+  OcpApimSubscriptionKey | "Content-Type",
+  never,
+  BasicResponseTypeWith401<ExtendedProfile>
+>;
+
+export type GetServicesByRecipientT = IGetApiRequestType<
+  {
+    readonly fiscalCode: string;
+  },
+  OcpApimSubscriptionKey,
+  never,
+  BasicResponseTypeWith401<Services>
+>;
+
+export type GetMessagesT = IGetApiRequestType<
+  {
+    readonly fiscalCode: string;
+  },
+  OcpApimSubscriptionKey,
+  never,
+  BasicResponseTypeWith401<Messages>
+>;
+
+export type GetMessageT = IGetApiRequestType<
+  {
+    readonly fiscalCode: string;
+    readonly id: string;
+  },
+  OcpApimSubscriptionKey,
+  never,
+  BasicResponseTypeWith401<MessageResponseWithContent>
+>;
+
+export type GetServicesT = IGetApiRequestType<
+  {},
+  OcpApimSubscriptionKey,
+  never,
+  BasicResponseTypeWith401<Services>
+>;
+
+export type GetServiceT = IGetApiRequestType<
+  {
+    readonly id: string;
+  },
+  OcpApimSubscriptionKey,
+  never,
+  BasicResponseTypeWith401<ServicePublic>
+>;
+
+export function APIClient(
+  baseUrl: string,
+  token: string,
+  // tslint:disable-next-line:no-any
+  fetchApi: typeof fetch = (nodeFetch as any) as typeof fetch
+): {
+  readonly createOrUpdateProfile: (
```

Comment:
you can use [TypeOfApiCall](https://github.com/teamdigitale/italia-ts-commons/blob/master/src/requests.ts#L228):

```ts
readonly createOrUpdateProfile: TypeOfApiCall<CreateOrUpdateProfileT>
```

Reply:
Ok, thank you. Refactored.